src/sinks/util/rusoto.rs

```diff
@@ -1,13 +1,97 @@
 use crate::dns::Resolver;
+use futures::{
+    future::{Future, FutureResult},
+    Poll,
+};
 use hyper::client::connect::HttpConnector;
 use hyper_openssl::{
     openssl::ssl::{SslConnector, SslMethod},
     HttpsConnector,
 };
-use rusoto_core::HttpClient;
+use rusoto_core::{CredentialsError, HttpClient, Region};
+use rusoto_credential::{
+    AutoRefreshingProvider, AutoRefreshingProviderFuture, AwsCredentials,
+    DefaultCredentialsProvider, DefaultCredentialsProviderFuture, ProvideAwsCredentials,
+    StaticProvider,
+};
+use rusoto_sts::{StsAssumeRoleSessionCredentialsProvider, StsClient};
+use snafu::{ResultExt, Snafu};
 
 pub type Client = HttpClient<HttpsConnector<HttpConnector<Resolver>>>;
 
+#[derive(Debug, Snafu)]
+enum RusotoError {
+    #[snafu(display("Invalid AWS credentials: {}", source))]
+    InvalidAWSCredentials { source: CredentialsError },
+}
+
+// A place-holder for the types of AWS credentials we support
+pub enum AwsCredentialsProvider {
+    Default(DefaultCredentialsProvider),
+    Role(AutoRefreshingProvider<StsAssumeRoleSessionCredentialsProvider>),
+    Static(StaticProvider),
+}
+
+impl AwsCredentialsProvider {
+    pub fn new(region: &Region, assume_role: Option<String>) -> crate::Result<Self> {
+        if let Some(role) = assume_role {
+            let sts = StsClient::new(region.clone());
+
+            let provider = StsAssumeRoleSessionCredentialsProvider::new(
+                sts,
+                role,
+                "default".to_owned(),
+                None,
+                None,
+                None,
+                None,
+            );
+
+            let creds = AutoRefreshingProvider::new(provider).context(InvalidAWSCredentials)?;
+            Ok(Self::Role(creds))
+        } else {
+            let creds = DefaultCredentialsProvider::new().context(InvalidAWSCredentials)?;
+            Ok(Self::Default(creds))
+        }
+    }
+
+    pub fn new_minimal<A: Into<String>, S: Into<String>>(access_key: A, secret_key: S) -> Self {
```

Comment:
Also curious if this should be called `new_static`?

Reply:
I thought about that, but the static provider has both `new_minimal` and a plain `new`, and I thought I'd reflect which one is being called, giving room for a `new_static` to provide the full parameters if needed.

Reply:
Ah ok that is fine!